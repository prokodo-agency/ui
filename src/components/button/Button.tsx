import { lazy, Suspense, type JSX } from 'react'

import ButtonServer from './Button.server'

import type { ButtonProps } from './Button.model'

/* ---------- preload chunk once ---------- */
const preload = () => import('./Button.lazy')
if (typeof window !== 'undefined') void preload()

/* ---------- typed lazy wrapper ---------- */
const LazyWrapper =
  typeof window !== 'undefined'
    ? lazy(() =>
        preload().then((m) => ({ default: m.default })), // <- adapter
      )
    : null

/* ---------------------------------------- */

export function Button({ priority = false, ...props }: ButtonProps): JSX.Element {
  const interactive =
    !!props.onClick || !!props.onKeyDown || !!props.redirect

  if (!interactive) return <ButtonServer {...props} />

  /* identical wrapper div on server & first paint */
  const Fallback = (
    <div data-island="button">
      <ButtonServer {...props} />
    </div>
  )

  return (
    <Suspense fallback={Fallback}>
      {LazyWrapper ? (
        <LazyWrapper {...props} priority={priority} />
      ) : (
        Fallback /* server branch renders the same HTML */
      )}
    </Suspense>
  )
}
