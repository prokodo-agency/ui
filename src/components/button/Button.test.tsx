import { expect } from "@jest/globals"

import { render } from "@/tests"

import { Button } from "./Button"

describe("The Button component", () => {
  it("should render Button", () => {
    const { container } = render(<Button title="Test title" />)
    expect(container.firstChild).toMatchSnapshot()
  })

  it("should render disabled Button", () => {
    const { container } = render(<Button disabled title="Test title" />)
    expect(container.firstChild).toMatchSnapshot()
  })

  it("should render loading Button", () => {
    const { container } = render(<Button loading title="Test title" />)
    expect(container.firstChild).toMatchSnapshot()
  })
})
