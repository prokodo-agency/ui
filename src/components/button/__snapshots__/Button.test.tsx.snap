// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Button component should render Button 1`] = `
<button
  aria-label="Test title"
  class="prokodo-Button prokodo-Button--has-variant-undefined"
  title="Test title"
  type="button"
>
  Test title
</button>
`;

exports[`The Button component should render disabled Button 1`] = `
<button
  aria-label="Test title"
  class="prokodo-Button prokodo-Button--has-variant-undefined prokodo-Button--is-disabled"
  disabled=""
  tabindex="-1"
  title="Test title"
  type="button"
>
  Test title
</button>
`;

exports[`The Button component should render loading Button 1`] = `
<button
  aria-label="Test title"
  class="prokodo-Button prokodo-Button--has-variant-undefined"
  title="Test title"
  type="button"
>
  Test title
</button>
`;
